File: ObjCGettingStarted/ObjCGettingStarted/ViewController.m
//
//  ViewController.m
//  ObjCGettingStarted
//
//  Created by JaesungMoon on 2018/04/13.
//  Copyright © 2018年 jaesung. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController
NSString *cellId = @"cellId";
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Courses";
    self.navigationController.navigationBar.prefersLargeTitles = YES;

    [self.tableView registerClass:UITableViewCell.class forCellReuseIdentifier:cellId];
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 10;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellId forIndexPath:indexPath];

    cell.textLabel.text = @"Course Name";
    return cell;
}

@end

File: SwiftThreading/SwiftThreading/ViewController.swift
//
//  ViewController.swift
//  SwiftThreading
//
//  Created by JaesungMoon on 2017/10/23.
//  Copyright © 2017年 JaesungMoon. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var tableView: UITableView!
    var tableData: [ModelData] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let start = Date()
        
        // DispatchQoS.userInteractive
        // DispatchQoS.userInitiated
        // DispatchQoS.background
        
        //async: your program doesn't wait for code to finish
        //sync: your program waits for code/task to finsh before countiuning to the next task
        
        
        //closure: Closures are self-contained blocks of functionality that cna be passed around and used in your code.
        DispatchQueue.global(qos: .userInteractive).async {
            Data.getData(completion: { (data) in
                self.tableData = data
                self.tableView.reloadData()
            })
            
        }
        
        let end = Date()
        print("Elapsed Time: \(end.timeIntervalSince(start))")
    }
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tableData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellId", for: indexPath)
        let data = tableData[indexPath.row]
        cell.textLabel?.text = data.name
        return cell
    }
}

